{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": []
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "hk8nSqciKNeA"
      },
      "outputs": [],
      "source": [
        "# evo_visualizer.py\n",
        "import sys\n",
        "# sys.path.insert(0, '/path/to/nsga-net')\n",
        "sys.path.insert(0, '/Users/zhichao.lu/Dropbox/2019/github/nsga-net')\n",
        "\n",
        "from graphviz import Digraph"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "import torch\n",
        "import torch.nn as nn\n",
        "from copy import copy\n",
        "from abc import ABC, abstractmethod\n",
        "from collections import OrderedDict"
      ],
      "metadata": {
        "id": "02xZyeFxLKkF"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "class Identity(nn.Module):\n",
        "    \"\"\"\n",
        "    Adding an identity allows us to keep things general in certain places.\n",
        "    \"\"\"\n",
        "\n",
        "    def __init__(self):\n",
        "        super(Identity, self).__init__()\n",
        "\n",
        "    def forward(self, x):\n",
        "        return "
      ],
      "metadata": {
        "id": "XSmOjPlILYdK"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "class PreactResidualNode(nn.Module):\n",
        "    \"\"\"\n",
        "    Basic computation unit.\n",
        "    Does batchnorm, relu, and convolution (in this order).\n",
        "    \"\"\"\n",
        "\n",
        "    def __init__(self, in_channels, out_channels, stride=1,\n",
        "                 kernel_size=3, padding=1, bias=False):\n",
        "        \"\"\"\n",
        "        Constructor.\n",
        "        Default arguments preserve dimensionality of input.\n",
        "        :param in_channels: input to the node.\n",
        "        :param out_channels: output channels from the node.\n",
        "        :param stride: stride of convolution, default 1.\n",
        "        :param kernel_size: size of convolution kernel, default 3.\n",
        "        :param padding: amount of zero padding, default 1.\n",
        "        :param bias: true to use bias, false to not.\n",
        "        \"\"\"\n",
        "        super(PreactResidualNode, self).__init__()\n",
        "\n",
        "        self.model = nn.Sequential(\n",
        "            nn.BatchNorm2d(in_channels),\n",
        "            nn.ReLU(inplace=True),\n",
        "            nn.Conv2d(in_channels, out_channels, kernel_size=kernel_size, stride=stride, padding=padding, bias=bias)\n",
        "        )\n",
        "\n",
        "    def forward(self, x):\n",
        "        \"\"\"\n",
        "        Apply forward propagation operation.\n",
        "        :param x: Variable, input.\n",
        "        :return: Variable.\n",
        "        \"\"\"\n",
        "        return self.model(x)\n"
      ],
      "metadata": {
        "id": "MoAZNlDALeWR"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "class ResidualPhase(nn.Module):\n",
        "    \"\"\"\n",
        "    Residual Genome phase.\n",
        "    \"\"\"\n",
        "\n",
        "    def __init__(self, gene, in_channels, out_channels, idx, preact=False):\n",
        "        \"\"\"\n",
        "        Constructor.\n",
        "        :param gene: list, element of genome describing connections in this phase.\n",
        "        :param in_channels: int, number of input channels.\n",
        "        :param out_channels: int, number of output channels.\n",
        "        :param idx: int, index in the network.\n",
        "        :param preact: should we use the preactivation scheme?\n",
        "        \"\"\"\n",
        "        super(ResidualPhase, self).__init__()\n",
        "\n",
        "        self.channel_flag = in_channels != out_channels  # Flag to tell us if we need to increase channel size.\n",
        "        self.first_conv = nn.Conv2d(in_channels, out_channels, kernel_size=1 if idx != 0 else 3, stride=1, bias=False)\n",
        "        self.dependency_graph = ResidualPhase.build_dependency_graph(gene)\n",
        "\n",
        "        if preact:\n",
        "            node_constructor = PreactResidualNode\n",
        "\n",
        "        else:\n",
        "            node_constructor = ResidualNode\n",
        "\n",
        "        nodes = []\n",
        "        for i in range(len(gene)):\n",
        "            if len(self.dependency_graph[i + 1]) > 0:\n",
        "                nodes.append(node_constructor(out_channels, out_channels))\n",
        "            else:\n",
        "                nodes.append(None)  # Module list will ignore NoneType.\n",
        "\n",
        "        self.nodes = nn.ModuleList(nodes)\n",
        "\n",
        "        #\n",
        "        # At this point, we know which nodes will be receiving input from where.\n",
        "        # So, we build the 1x1 convolutions that will deal with the depth-wise concatenations.\n",
        "        #\n",
        "        conv1x1s = [Identity()] + [Identity() for _ in range(max(self.dependency_graph.keys()))]\n",
        "        for node_idx, dependencies in self.dependency_graph.items():\n",
        "            if len(dependencies) > 1:\n",
        "                conv1x1s[node_idx] = \\\n",
        "                    nn.Conv2d(len(dependencies) * out_channels, out_channels, kernel_size=1, bias=False)\n",
        "\n",
        "        self.processors = nn.ModuleList(conv1x1s)\n",
        "        self.out = nn.Sequential(\n",
        "            nn.BatchNorm2d(out_channels),\n",
        "            nn.ReLU(inplace=True)\n",
        "        )\n",
        "\n",
        "    @staticmethod\n",
        "    def build_dependency_graph(gene):\n",
        "        \"\"\"\n",
        "        Build a graph describing the connections of a phase.\n",
        "        \"Repairs\" made are as follows:\n",
        "            - If a node has no input, but gives output, connect it to the input node (index 0 in outputs).\n",
        "            - If a node has input, but no output, connect it to the output node (value returned from forward method).\n",
        "        :param gene: gene describing the phase connections.\n",
        "        :return: dict\n",
        "        \"\"\"\n",
        "        graph = {}\n",
        "        residual = gene[-1][0] == 1\n",
        "\n",
        "        # First pass, build the graph without repairs.\n",
        "        graph[1] = []\n",
        "        for i in range(len(gene) - 1):\n",
        "            graph[i + 2] = [j + 1 for j in range(len(gene[i])) if gene[i][j] == 1]\n",
        "\n",
        "        graph[len(gene) + 1] = [0] if residual else []\n",
        "\n",
        "        # Determine which nodes, if any, have no inputs and/or outputs.\n",
        "        no_inputs = []\n",
        "        no_outputs = []\n",
        "        for i in range(1, len(gene) + 1):\n",
        "            if len(graph[i]) == 0:\n",
        "                no_inputs.append(i)\n",
        "\n",
        "            has_output = False\n",
        "            for j in range(i + 1, len(gene) + 2):\n",
        "                if i in graph[j]:\n",
        "                    has_output = True\n",
        "                    break\n",
        "\n",
        "            if not has_output:\n",
        "                no_outputs.append(i)\n",
        "\n",
        "        for node in no_outputs:\n",
        "            if node not in no_inputs:\n",
        "                # No outputs, but has inputs. Connect to output node.\n",
        "                graph[len(gene) + 1].append(node)\n",
        "\n",
        "        for node in no_inputs:\n",
        "            if node not in no_outputs:\n",
        "                # No inputs, but has outputs. Connect to input node.\n",
        "                graph[node].append(0)\n",
        "\n",
        "        return graph\n",
        "\n",
        "    def forward(self, x):\n",
        "        if self.channel_flag:\n",
        "            x = self.first_conv(x)\n",
        "\n",
        "        outputs = [x]\n",
        "\n",
        "        for i in range(1, len(self.nodes) + 1):\n",
        "            if not self.dependency_graph[i]:  # Empty list, no outputs to give.\n",
        "                outputs.append(None)\n",
        "\n",
        "            else:\n",
        "                outputs.append(self.nodes[i - 1](self.process_dependencies(i, outputs)))\n",
        "\n",
        "        return self.out(self.process_dependencies(len(self.nodes) + 1, outputs))\n",
        "\n",
        "    def process_dependencies(self, node_idx, outputs):\n",
        "        \"\"\"\n",
        "        Process dependencies with a depth-wise concatenation and\n",
        "        :param node_idx: int,\n",
        "        :param outputs: list, current outputs\n",
        "        :return: Variable\n",
        "        \"\"\"\n",
        "        return self.processors[node_idx](torch.cat([outputs[i] for i in self.dependency_graph[node_idx]], dim=1))\n",
        "\n",
        "\n",
        "class ResidualNode(nn.Module):\n",
        "    \"\"\"\n",
        "    Basic computation unit.\n",
        "    Does convolution, batchnorm, and relu (in this order).\n",
        "    \"\"\"\n",
        "\n",
        "    def __init__(self, in_channels, out_channels, stride=1,\n",
        "                 kernel_size=3, padding=1, bias=False):\n",
        "        \"\"\"\n",
        "        Constructor.\n",
        "        Default arguments preserve dimensionality of input.\n",
        "        :param in_channels: input to the node.\n",
        "        :param out_channels: output channels from the node.\n",
        "        :param stride: stride of convolution, default 1.\n",
        "        :param kernel_size: size of convolution kernel, default 3.\n",
        "        :param padding: amount of zero padding, default 1.\n",
        "        :param bias: true to use bias, false to not.\n",
        "        \"\"\"\n",
        "        super(ResidualNode, self).__init__()\n",
        "\n",
        "        self.model = nn.Sequential(\n",
        "            nn.Conv2d(in_channels, out_channels, kernel_size=kernel_size, stride=stride, padding=padding, bias=bias),\n",
        "            nn.BatchNorm2d(out_channels),\n",
        "            nn.ReLU(inplace=True)\n",
        "        )\n",
        "\n",
        "    def forward(self, x):\n",
        "        \"\"\"\n",
        "        Apply forward propagation operation.\n",
        "        :param x: Variable, input.\n",
        "        :return: Variable.\n",
        "        \"\"\"\n",
        "        return self.model(x)\n"
      ],
      "metadata": {
        "id": "Gm0FlwyyKpBG"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "def get_graph_function(type):\n",
        "    \"\"\"\n",
        "    Get the appropriate function to build a dependency graph.\n",
        "    :param type:\n",
        "    :return: callable\n",
        "    \"\"\"\n",
        "    if type == \"residual\":\n",
        "        return ResidualPhase.build_dependency_graph\n",
        "\n",
        "    raise NotImplementedError(\"Genome type {} not supported.\".format(type))"
      ],
      "metadata": {
        "id": "w0doXVLdLvxf"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "def make_dot_phase(phase, rankdir=\"UD\", format=\"pdf\", title=None, filename=\"genome\", type=\"residual\"):\n",
        "    \"\"\"\n",
        "    Visualize a single phase.\n",
        "    :param genome: list of lists.\n",
        "    :param phase: int, index of phase to visualize.\n",
        "    :param rankdir: direction graph is oriented \"UD\"=Vertical, \"LR\"=horizontal.\n",
        "    :param format: output file format, jpg, png, etc.\n",
        "    :param title: title of graph.\n",
        "    :param filename: filename of graph.\n",
        "    :param type: string, what kind of decoder should we use.\n",
        "    :return: graphviz dot object.\n",
        "    :return:\n",
        "    \"\"\"\n",
        "    node_color = \"lightblue\"\n",
        "    conv1x1_color = \"white\"\n",
        "    sum_color = \"green4\"\n",
        "    pool_color = \"orange\"\n",
        "    phase_background_color = \"gray\"\n",
        "    fc_color = \"gray\"\n",
        "    node_shape = \"circle\"\n",
        "    conv1x1_shape = \"doublecircle\"\n",
        "\n",
        "    gene = phase\n",
        "    graph_function = get_graph_function(type)\n",
        "    graph = graph_function(gene)\n",
        "\n",
        "    nodes = [(\"node_0\", ' ')] + [(\"node_\" + str(j + 1), ' ') for j in range(len(gene) + 1)]\n",
        "    edges = []\n",
        "\n",
        "    for sink, dependencies in graph.items():\n",
        "        for source in dependencies:\n",
        "            edges.append((nodes[source][0], nodes[sink][0]))\n",
        "\n",
        "    node_attr = dict(style='filled',\n",
        "                     shape='box',\n",
        "                     align='left',\n",
        "                     fontsize='12',\n",
        "                     ranksep='0.1',\n",
        "                     height='0.2')\n",
        "    dot = Digraph(format=format, filename=filename+'.gv', node_attr=node_attr, graph_attr=dict(size=\"12,12\"))\n",
        "    dot.attr(rankdir=rankdir)\n",
        "\n",
        "    if title:\n",
        "        dot.attr(label=title+\"\\n\\n\")\n",
        "        dot.attr(labelloc='t')\n",
        "\n",
        "    dot.node(nodes[0][0], nodes[0][1], fillcolor=conv1x1_color, shape=conv1x1_shape)\n",
        "\n",
        "    with dot.subgraph(name=\"cluster\") as p:\n",
        "        p.attr(fillcolor=phase_background_color, label='', fontcolor=\"black\", style=\"filled\")\n",
        "\n",
        "        for i in range(1, len(nodes) - 1):\n",
        "            if len(graph[i]) != 0:\n",
        "                p.node(nodes[i][0], nodes[i][1], fillcolor=node_color, shape=node_shape)\n",
        "\n",
        "    dot.node(nodes[-1][0], nodes[-1][1], fillcolor=sum_color, shape=node_shape)\n",
        "\n",
        "    # Add edges.\n",
        "    for edge in edges:\n",
        "        dot.edge(*edge)\n",
        "\n",
        "    dot.attr(dpi=\"300\")\n",
        "\n",
        "    return "
      ],
      "metadata": {
        "id": "24No79YtLyeh"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "def make_dot_genome(genome, rankdir=\"UD\", format=\"pdf\", title=None, filename=\"genome\", type=\"residual\"):\n",
        "    \"\"\"\n",
        "    Graphviz representation of network created by genome.\n",
        "    :param genome: list of lists.\n",
        "    :param rankdir: direction graph is oriented \"UD\"=Vertical, \"LR\"=horizontal.\n",
        "    :param format: output file format, jpg, png, etc.\n",
        "    :param title: title of graph.\n",
        "    :param filename: filename of graph.\n",
        "    :param type: string, what kind of decoder should we use.\n",
        "    :return: graphviz dot object.\n",
        "    \"\"\"\n",
        "    node_color = \"lightblue\"\n",
        "    conv1x1_color = \"white\"\n",
        "    sum_color = \"green4\"\n",
        "    pool_color = \"orange\"\n",
        "    phase_background_color = \"gray\"\n",
        "    fc_color = \"gray\"\n",
        "\n",
        "    node_shape = \"circle\"\n",
        "    conv1x1_shape = \"doublecircle\"\n",
        "\n",
        "    structure = []\n",
        "\n",
        "    # Build node ids and names to make building graph easier.\n",
        "    for i, gene in enumerate(genome):\n",
        "        all_zeros = sum([sum(t) for t in gene[:-1]]) == 0\n",
        "\n",
        "        if all_zeros:\n",
        "            continue  # Skip everything is a gene is all zeros.\n",
        "\n",
        "        prefix = \"gene_\" + str(i)\n",
        "        phase = (\"cluster_\" + str(i + 1), \"Phase \" + str(i + 1))\n",
        "\n",
        "        nodes = [(prefix + \"_node_0\", ' ')] \\\n",
        "            + [(prefix + \"_node_\" + str(j + 1), ' ') for j in range(len(gene) + 1)]\n",
        "\n",
        "        pool = (prefix + \"_pool\", \"Pooling\")\n",
        "\n",
        "\n",
        "        edges = []\n",
        "        graph_function = get_graph_function(type)\n",
        "        graph = graph_function(gene)\n",
        "\n",
        "        for sink, dependencies in graph.items():\n",
        "            for source in dependencies:\n",
        "                edges.append((nodes[source][0], nodes[sink][0]))\n",
        "\n",
        "        structure.append(\n",
        "            {\n",
        "                \"nodes\": nodes,\n",
        "                \"edges\": edges,\n",
        "                \"pool\": pool,\n",
        "                \"phase\": phase,\n",
        "                \"all_zeros\": all_zeros,\n",
        "                \"graph\": graph\n",
        "            }\n",
        "        )\n",
        "\n",
        "    final_pool = structure[-1][\"pool\"]\n",
        "    new_pool = (final_pool[0], \"Avg. Pooling\")\n",
        "    structure[-1][\"pool\"] = new_pool\n",
        "\n",
        "    node_attr = dict(style='filled',\n",
        "                     shape='box',\n",
        "                     align='left',\n",
        "                     fontsize='12',\n",
        "                     ranksep='0.1',\n",
        "                     height='0.2')\n",
        "    dot = Digraph(format=format, filename=filename+'.gv', node_attr=node_attr, graph_attr=dict(size=\"12,12\"))\n",
        "    dot.attr(rankdir=rankdir)\n",
        "\n",
        "    if title:\n",
        "        dot.attr(label=title+\"\\n\\n\")\n",
        "        dot.attr(labelloc='t')\n",
        "\n",
        "    dot.node(str(\"input\"), \"Input\")\n",
        "\n",
        "    for j, struct in enumerate(structure):\n",
        "        nodes = struct['nodes']\n",
        "        edges = struct['edges']\n",
        "        phase = struct['phase']\n",
        "        pool = struct['pool']\n",
        "        graph = struct['graph']\n",
        "        all_zeros = struct['all_zeros']\n",
        "\n",
        "        # Add nodes.\n",
        "        dot.node(nodes[0][0], nodes[0][1], fillcolor=conv1x1_color, shape=conv1x1_shape)\n",
        "\n",
        "        if j > 0:\n",
        "            dot.edge(structure[j - 1]['pool'][0], nodes[0][0])\n",
        "\n",
        "        if not all_zeros:\n",
        "            with dot.subgraph(name=phase[0]) as p:\n",
        "                p.attr(fillcolor=phase_background_color, label='', fontcolor=\"black\", style=\"filled\")\n",
        "\n",
        "                for i in range(1, len(nodes) - 1):\n",
        "                    if len(graph[i]) != 0:\n",
        "                        p.node(nodes[i][0], nodes[i][1], fillcolor=node_color, shape=node_shape)\n",
        "\n",
        "        dot.node(nodes[-1][0], nodes[-1][1], fillcolor=sum_color, shape=node_shape)\n",
        "        dot.node(*pool, fillcolor=pool_color)\n",
        "\n",
        "        # Add edges.\n",
        "        for edge in edges:\n",
        "            dot.edge(*edge)\n",
        "\n",
        "        dot.edge(nodes[-1][0], pool[0])\n",
        "\n",
        "    dot.edge(\"input\", structure[0]['nodes'][0][0])\n",
        "\n",
        "    dot.node(\"linear\", \"Linear\", fillcolor=fc_color)\n",
        "    dot.edge(structure[-1]['pool'][0], \"linear\")\n",
        "\n",
        "    # TODO: Add legend (?)\n",
        "\n",
        "    return dot"
      ],
      "metadata": {
        "id": "7g8yF36sL2Xa"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "from graphviz import Graph"
      ],
      "metadata": {
        "id": "DREesIHBPpan"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "import graphviz"
      ],
      "metadata": {
        "id": "MMLuFDXtP16C"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "import numpy as np"
      ],
      "metadata": {
        "id": "haH15V2ERzRC"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "def phase_dencode(phase_bit_string):\n",
        "    n = int(np.sqrt(2 * len(phase_bit_string) - 7/4) - 1/2)\n",
        "    genome = []\n",
        "    for i in range(n):\n",
        "        operator = []\n",
        "        for j in range(i + 1):\n",
        "            operator.append(phase_bit_string[int(i * (i + 1) / 2 + j)])\n",
        "        genome.append(operator)\n",
        "    genome.append([phase_bit_string[-1]])\n",
        "    return genome"
      ],
      "metadata": {
        "id": "ERsnSBd-RyLK"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "Genome = [[1, 1, 0, 0, 0, 1, 1, 0, 1, 0, 1, 0, 0, 0, 0, 1], [0, 1, 1, 1, 1, 1, 0, 1, 1, 1, 0, 1, 0, 1, 1, 0], [1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 1, 0, 0, 1, 0, 0]]"
      ],
      "metadata": {
        "id": "rje2LEkSR3Kh"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "genotype_name = \"NSGANet\""
      ],
      "metadata": {
        "id": "f1hhbaFXMgoD"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "NSGANet = [[[1], [1, 0], [1, 1, 1], [1]], [[0], [0, 1], [0, 0, 1], [0]], [[1], [0, 0], [1, 0, 1], [1]]]\n",
        "d = make_dot_genome(NSGANet, title=\"{}\".format(genotype_name), filename=\"macro_network\")\n",
        "d.view()\n",
        "with open(\"/content/macro_network.gv\") as f:\n",
        "    dot_graph = f.read()\n",
        "graphviz.Source(dot_graph)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 1000
        },
        "id": "zCIlSCnsMM1R",
        "outputId": "07a54023-772d-4644-fa7e-b064fa62b448"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "<graphviz.files.Source at 0x7ff73926cb80>"
            ],
            "image/svg+xml": "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n<!-- Generated by graphviz version 2.40.1 (20161225.0304)\n -->\n<!-- Title: %3 Pages: 1 -->\n<svg width=\"108pt\" height=\"864pt\"\n viewBox=\"0.00 0.00 108.02 864.00\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n<g id=\"graph0\" class=\"graph\" transform=\"scale(.6969 .6969) rotate(0) translate(4 1235.7788)\">\n<title>%3</title>\n<polygon fill=\"#ffffff\" stroke=\"transparent\" points=\"-4,4 -4,-1235.7788 151,-1235.7788 151,4 -4,4\"/>\n<text text-anchor=\"middle\" x=\"73.5\" y=\"-1216.5788\" font-family=\"Times,serif\" font-size=\"14.00\" fill=\"#000000\">NSGANet</text>\n<g id=\"clust1\" class=\"cluster\">\n<title>cluster_1</title>\n<polygon fill=\"#c0c0c0\" stroke=\"#000000\" points=\"8,-892.9848 8,-1070.0803 122,-1070.0803 122,-892.9848 8,-892.9848\"/>\n</g>\n<g id=\"clust2\" class=\"cluster\">\n<title>cluster_2</title>\n<polygon fill=\"#c0c0c0\" stroke=\"#000000\" points=\"71,-499.4924 71,-676.5879 117,-676.5879 117,-499.4924 71,-499.4924\"/>\n</g>\n<g id=\"clust3\" class=\"cluster\">\n<title>cluster_3</title>\n<polygon fill=\"#c0c0c0\" stroke=\"#000000\" points=\"23,-171.6985 23,-283.0955 117,-283.0955 117,-171.6985 23,-171.6985\"/>\n</g>\n<!-- input -->\n<g id=\"node1\" class=\"node\">\n<title>input</title>\n<polygon fill=\"#d3d3d3\" stroke=\"#000000\" points=\"147,-1192.7788 93,-1192.7788 93,-1171.7788 147,-1171.7788 147,-1192.7788\"/>\n<text text-anchor=\"middle\" x=\"120\" y=\"-1179.1788\" font-family=\"Times,serif\" font-size=\"12.00\" fill=\"#000000\">Input</text>\n</g>\n<!-- gene_0_node_0 -->\n<g id=\"node2\" class=\"node\">\n<title>gene_0_node_0</title>\n<ellipse fill=\"#ffffff\" stroke=\"#000000\" cx=\"120\" cy=\"-1116.9295\" rx=\"14.7314\" ry=\"14.7314\"/>\n<ellipse fill=\"none\" stroke=\"#000000\" cx=\"120\" cy=\"-1116.9295\" rx=\"18.6997\" ry=\"18.6997\"/>\n<text text-anchor=\"middle\" x=\"120\" y=\"-1113.8295\" font-family=\"Times,serif\" font-size=\"12.00\" fill=\"#000000\"> </text>\n</g>\n<!-- input&#45;&gt;gene_0_node_0 -->\n<g id=\"edge26\" class=\"edge\">\n<title>input&#45;&gt;gene_0_node_0</title>\n<path fill=\"none\" stroke=\"#000000\" d=\"M120,-1171.4877C120,-1164.5568 120,-1155.1946 120,-1146.1833\"/>\n<polygon fill=\"#000000\" stroke=\"#000000\" points=\"123.5001,-1145.9327 120,-1135.9328 116.5001,-1145.9328 123.5001,-1145.9327\"/>\n</g>\n<!-- gene_0_node_1 -->\n<g id=\"node3\" class=\"node\">\n<title>gene_0_node_1</title>\n<ellipse fill=\"#add8e6\" stroke=\"#000000\" cx=\"89\" cy=\"-1047.2311\" rx=\"14.7\" ry=\"14.7\"/>\n<text text-anchor=\"middle\" x=\"89\" y=\"-1044.1311\" font-family=\"Times,serif\" font-size=\"12.00\" fill=\"#000000\"> </text>\n</g>\n<!-- gene_0_node_0&#45;&gt;gene_0_node_1 -->\n<g id=\"edge1\" class=\"edge\">\n<title>gene_0_node_0&#45;&gt;gene_0_node_1</title>\n<path fill=\"none\" stroke=\"#000000\" d=\"M112.3371,-1099.7007C108.3982,-1090.8447 103.5429,-1079.9284 99.2829,-1070.3504\"/>\n<polygon fill=\"#000000\" stroke=\"#000000\" points=\"102.3842,-1068.7107 95.1223,-1060.996 95.9883,-1071.5554 102.3842,-1068.7107\"/>\n</g>\n<!-- gene_0_node_5 -->\n<g id=\"node7\" class=\"node\">\n<title>gene_0_node_5</title>\n<ellipse fill=\"#008b00\" stroke=\"#000000\" cx=\"94\" cy=\"-850.1356\" rx=\"14.7\" ry=\"14.7\"/>\n<text text-anchor=\"middle\" x=\"94\" y=\"-847.0356\" font-family=\"Times,serif\" font-size=\"12.00\" fill=\"#000000\"> </text>\n</g>\n<!-- gene_0_node_0&#45;&gt;gene_0_node_5 -->\n<g id=\"edge7\" class=\"edge\">\n<title>gene_0_node_0&#45;&gt;gene_0_node_5</title>\n<path fill=\"none\" stroke=\"#000000\" d=\"M124.5904,-1098.6114C133.4964,-1059.5903 150.2161,-965.882 126,-892.9848 123.0126,-883.9919 117.3209,-875.3589 111.5561,-868.2281\"/>\n<polygon fill=\"#000000\" stroke=\"#000000\" points=\"114.1251,-865.8488 104.9128,-860.6158 108.8511,-870.4515 114.1251,-865.8488\"/>\n</g>\n<!-- gene_0_node_2 -->\n<g id=\"node4\" class=\"node\">\n<title>gene_0_node_2</title>\n<ellipse fill=\"#add8e6\" stroke=\"#000000\" cx=\"31\" cy=\"-981.5326\" rx=\"14.7\" ry=\"14.7\"/>\n<text text-anchor=\"middle\" x=\"31\" y=\"-978.4326\" font-family=\"Times,serif\" font-size=\"12.00\" fill=\"#000000\"> </text>\n</g>\n<!-- gene_0_node_1&#45;&gt;gene_0_node_2 -->\n<g id=\"edge2\" class=\"edge\">\n<title>gene_0_node_1&#45;&gt;gene_0_node_2</title>\n<path fill=\"none\" stroke=\"#000000\" d=\"M79.1611,-1036.0862C70.6307,-1026.4236 58.0967,-1012.2259 47.9538,-1000.7366\"/>\n<polygon fill=\"#000000\" stroke=\"#000000\" points=\"50.3242,-998.1332 41.0821,-992.9529 45.0765,-1002.766 50.3242,-998.1332\"/>\n</g>\n<!-- gene_0_node_3 -->\n<g id=\"node5\" class=\"node\">\n<title>gene_0_node_3</title>\n<ellipse fill=\"#add8e6\" stroke=\"#000000\" cx=\"79\" cy=\"-981.5326\" rx=\"14.7\" ry=\"14.7\"/>\n<text text-anchor=\"middle\" x=\"79\" y=\"-978.4326\" font-family=\"Times,serif\" font-size=\"12.00\" fill=\"#000000\"> </text>\n</g>\n<!-- gene_0_node_1&#45;&gt;gene_0_node_3 -->\n<g id=\"edge3\" class=\"edge\">\n<title>gene_0_node_1&#45;&gt;gene_0_node_3</title>\n<path fill=\"none\" stroke=\"#000000\" d=\"M86.7308,-1032.3231C85.5674,-1024.6791 84.1193,-1015.1654 82.792,-1006.4454\"/>\n<polygon fill=\"#000000\" stroke=\"#000000\" points=\"86.2252,-1005.741 81.2602,-996.3816 79.3049,-1006.7944 86.2252,-1005.741\"/>\n</g>\n<!-- gene_0_node_4 -->\n<g id=\"node6\" class=\"node\">\n<title>gene_0_node_4</title>\n<ellipse fill=\"#add8e6\" stroke=\"#000000\" cx=\"84\" cy=\"-915.8341\" rx=\"14.7\" ry=\"14.7\"/>\n<text text-anchor=\"middle\" x=\"84\" y=\"-912.7341\" font-family=\"Times,serif\" font-size=\"12.00\" fill=\"#000000\"> </text>\n</g>\n<!-- gene_0_node_1&#45;&gt;gene_0_node_4 -->\n<g id=\"edge4\" class=\"edge\">\n<title>gene_0_node_1&#45;&gt;gene_0_node_4</title>\n<path fill=\"none\" stroke=\"#000000\" d=\"M94.4283,-1033.2517C99.9734,-1017.1791 107.2155,-990.0088 103,-966.6833 101.2894,-957.2182 97.9228,-947.2178 94.4925,-938.6135\"/>\n<polygon fill=\"#000000\" stroke=\"#000000\" points=\"97.657,-937.1108 90.5318,-929.2698 91.2121,-939.8428 97.657,-937.1108\"/>\n</g>\n<!-- gene_0_node_2&#45;&gt;gene_0_node_4 -->\n<g id=\"edge5\" class=\"edge\">\n<title>gene_0_node_2&#45;&gt;gene_0_node_4</title>\n<path fill=\"none\" stroke=\"#000000\" d=\"M40.4753,-969.787C48.1628,-960.2577 59.165,-946.6194 68.1841,-935.4394\"/>\n<polygon fill=\"#000000\" stroke=\"#000000\" points=\"71.052,-937.4587 74.6067,-927.478 65.6038,-933.0635 71.052,-937.4587\"/>\n</g>\n<!-- gene_0_node_3&#45;&gt;gene_0_node_4 -->\n<g id=\"edge6\" class=\"edge\">\n<title>gene_0_node_3&#45;&gt;gene_0_node_4</title>\n<path fill=\"none\" stroke=\"#000000\" d=\"M80.1346,-966.6246C80.7163,-958.9806 81.4404,-949.467 82.104,-940.747\"/>\n<polygon fill=\"#000000\" stroke=\"#000000\" points=\"85.6009,-940.9199 82.8699,-930.6831 78.6211,-940.3886 85.6009,-940.9199\"/>\n</g>\n<!-- gene_0_node_4&#45;&gt;gene_0_node_5 -->\n<g id=\"edge8\" class=\"edge\">\n<title>gene_0_node_4&#45;&gt;gene_0_node_5</title>\n<path fill=\"none\" stroke=\"#000000\" d=\"M86.2692,-900.9261C87.4326,-893.2821 88.8807,-883.7685 90.208,-875.0485\"/>\n<polygon fill=\"#000000\" stroke=\"#000000\" points=\"93.6951,-875.3974 91.7398,-864.9846 86.7748,-874.344 93.6951,-875.3974\"/>\n</g>\n<!-- gene_0_pool -->\n<g id=\"node8\" class=\"node\">\n<title>gene_0_pool</title>\n<polygon fill=\"#ffa500\" stroke=\"#000000\" points=\"121,-799.2864 67,-799.2864 67,-778.2864 121,-778.2864 121,-799.2864\"/>\n<text text-anchor=\"middle\" x=\"94\" y=\"-785.6864\" font-family=\"Times,serif\" font-size=\"12.00\" fill=\"#000000\">Pooling</text>\n</g>\n<!-- gene_0_node_5&#45;&gt;gene_0_pool -->\n<g id=\"edge9\" class=\"edge\">\n<title>gene_0_node_5&#45;&gt;gene_0_pool</title>\n<path fill=\"none\" stroke=\"#000000\" d=\"M94,-835.2847C94,-827.5917 94,-818.0701 94,-809.6506\"/>\n<polygon fill=\"#000000\" stroke=\"#000000\" points=\"97.5001,-809.4638 94,-799.4638 90.5001,-809.4639 97.5001,-809.4638\"/>\n</g>\n<!-- gene_1_node_0 -->\n<g id=\"node9\" class=\"node\">\n<title>gene_1_node_0</title>\n<ellipse fill=\"#ffffff\" stroke=\"#000000\" cx=\"94\" cy=\"-723.4371\" rx=\"14.7314\" ry=\"14.7314\"/>\n<ellipse fill=\"none\" stroke=\"#000000\" cx=\"94\" cy=\"-723.4371\" rx=\"18.6997\" ry=\"18.6997\"/>\n<text text-anchor=\"middle\" x=\"94\" y=\"-720.3371\" font-family=\"Times,serif\" font-size=\"12.00\" fill=\"#000000\"> </text>\n</g>\n<!-- gene_0_pool&#45;&gt;gene_1_node_0 -->\n<g id=\"edge10\" class=\"edge\">\n<title>gene_0_pool&#45;&gt;gene_1_node_0</title>\n<path fill=\"none\" stroke=\"#000000\" d=\"M94,-777.9953C94,-771.0643 94,-761.7022 94,-752.6909\"/>\n<polygon fill=\"#000000\" stroke=\"#000000\" points=\"97.5001,-752.4403 94,-742.4403 90.5001,-752.4404 97.5001,-752.4403\"/>\n</g>\n<!-- gene_1_node_2 -->\n<g id=\"node10\" class=\"node\">\n<title>gene_1_node_2</title>\n<ellipse fill=\"#add8e6\" stroke=\"#000000\" cx=\"94\" cy=\"-653.7386\" rx=\"14.7\" ry=\"14.7\"/>\n<text text-anchor=\"middle\" x=\"94\" y=\"-650.6386\" font-family=\"Times,serif\" font-size=\"12.00\" fill=\"#000000\"> </text>\n</g>\n<!-- gene_1_node_0&#45;&gt;gene_1_node_2 -->\n<g id=\"edge11\" class=\"edge\">\n<title>gene_1_node_0&#45;&gt;gene_1_node_2</title>\n<path fill=\"none\" stroke=\"#000000\" d=\"M94,-704.3917C94,-696.5699 94,-687.4005 94,-679.0288\"/>\n<polygon fill=\"#000000\" stroke=\"#000000\" points=\"97.5001,-679.0127 94,-669.0127 90.5001,-679.0128 97.5001,-679.0127\"/>\n</g>\n<!-- gene_1_node_3 -->\n<g id=\"node11\" class=\"node\">\n<title>gene_1_node_3</title>\n<ellipse fill=\"#add8e6\" stroke=\"#000000\" cx=\"94\" cy=\"-588.0402\" rx=\"14.7\" ry=\"14.7\"/>\n<text text-anchor=\"middle\" x=\"94\" y=\"-584.9402\" font-family=\"Times,serif\" font-size=\"12.00\" fill=\"#000000\"> </text>\n</g>\n<!-- gene_1_node_2&#45;&gt;gene_1_node_3 -->\n<g id=\"edge12\" class=\"edge\">\n<title>gene_1_node_2&#45;&gt;gene_1_node_3</title>\n<path fill=\"none\" stroke=\"#000000\" d=\"M94,-638.8306C94,-631.3771 94,-622.146 94,-613.6063\"/>\n<polygon fill=\"#000000\" stroke=\"#000000\" points=\"97.5001,-613.345 94,-603.3451 90.5001,-613.3451 97.5001,-613.345\"/>\n</g>\n<!-- gene_1_node_4 -->\n<g id=\"node12\" class=\"node\">\n<title>gene_1_node_4</title>\n<ellipse fill=\"#add8e6\" stroke=\"#000000\" cx=\"94\" cy=\"-522.3417\" rx=\"14.7\" ry=\"14.7\"/>\n<text text-anchor=\"middle\" x=\"94\" y=\"-519.2417\" font-family=\"Times,serif\" font-size=\"12.00\" fill=\"#000000\"> </text>\n</g>\n<!-- gene_1_node_3&#45;&gt;gene_1_node_4 -->\n<g id=\"edge13\" class=\"edge\">\n<title>gene_1_node_3&#45;&gt;gene_1_node_4</title>\n<path fill=\"none\" stroke=\"#000000\" d=\"M94,-573.1321C94,-565.6787 94,-556.4475 94,-547.9078\"/>\n<polygon fill=\"#000000\" stroke=\"#000000\" points=\"97.5001,-547.6466 94,-537.6466 90.5001,-547.6466 97.5001,-547.6466\"/>\n</g>\n<!-- gene_1_node_5 -->\n<g id=\"node13\" class=\"node\">\n<title>gene_1_node_5</title>\n<ellipse fill=\"#008b00\" stroke=\"#000000\" cx=\"94\" cy=\"-456.6432\" rx=\"14.7\" ry=\"14.7\"/>\n<text text-anchor=\"middle\" x=\"94\" y=\"-453.5432\" font-family=\"Times,serif\" font-size=\"12.00\" fill=\"#000000\"> </text>\n</g>\n<!-- gene_1_node_4&#45;&gt;gene_1_node_5 -->\n<g id=\"edge14\" class=\"edge\">\n<title>gene_1_node_4&#45;&gt;gene_1_node_5</title>\n<path fill=\"none\" stroke=\"#000000\" d=\"M94,-507.4337C94,-499.9802 94,-490.749 94,-482.2094\"/>\n<polygon fill=\"#000000\" stroke=\"#000000\" points=\"97.5001,-481.9481 94,-471.9481 90.5001,-481.9481 97.5001,-481.9481\"/>\n</g>\n<!-- gene_1_pool -->\n<g id=\"node14\" class=\"node\">\n<title>gene_1_pool</title>\n<polygon fill=\"#ffa500\" stroke=\"#000000\" points=\"121,-405.7939 67,-405.7939 67,-384.7939 121,-384.7939 121,-405.7939\"/>\n<text text-anchor=\"middle\" x=\"94\" y=\"-392.1939\" font-family=\"Times,serif\" font-size=\"12.00\" fill=\"#000000\">Pooling</text>\n</g>\n<!-- gene_1_node_5&#45;&gt;gene_1_pool -->\n<g id=\"edge15\" class=\"edge\">\n<title>gene_1_node_5&#45;&gt;gene_1_pool</title>\n<path fill=\"none\" stroke=\"#000000\" d=\"M94,-441.7922C94,-434.0992 94,-424.5777 94,-416.1582\"/>\n<polygon fill=\"#000000\" stroke=\"#000000\" points=\"97.5001,-415.9714 94,-405.9714 90.5001,-415.9714 97.5001,-415.9714\"/>\n</g>\n<!-- gene_2_node_0 -->\n<g id=\"node15\" class=\"node\">\n<title>gene_2_node_0</title>\n<ellipse fill=\"#ffffff\" stroke=\"#000000\" cx=\"94\" cy=\"-329.9447\" rx=\"14.7314\" ry=\"14.7314\"/>\n<ellipse fill=\"none\" stroke=\"#000000\" cx=\"94\" cy=\"-329.9447\" rx=\"18.6997\" ry=\"18.6997\"/>\n<text text-anchor=\"middle\" x=\"94\" y=\"-326.8447\" font-family=\"Times,serif\" font-size=\"12.00\" fill=\"#000000\"> </text>\n</g>\n<!-- gene_1_pool&#45;&gt;gene_2_node_0 -->\n<g id=\"edge16\" class=\"edge\">\n<title>gene_1_pool&#45;&gt;gene_2_node_0</title>\n<path fill=\"none\" stroke=\"#000000\" d=\"M94,-384.5028C94,-377.5719 94,-368.2097 94,-359.1985\"/>\n<polygon fill=\"#000000\" stroke=\"#000000\" points=\"97.5001,-358.9479 94,-348.9479 90.5001,-358.9479 97.5001,-358.9479\"/>\n</g>\n<!-- gene_2_node_1 -->\n<g id=\"node16\" class=\"node\">\n<title>gene_2_node_1</title>\n<ellipse fill=\"#add8e6\" stroke=\"#000000\" cx=\"94\" cy=\"-260.2462\" rx=\"14.7\" ry=\"14.7\"/>\n<text text-anchor=\"middle\" x=\"94\" y=\"-257.1462\" font-family=\"Times,serif\" font-size=\"12.00\" fill=\"#000000\"> </text>\n</g>\n<!-- gene_2_node_0&#45;&gt;gene_2_node_1 -->\n<g id=\"edge17\" class=\"edge\">\n<title>gene_2_node_0&#45;&gt;gene_2_node_1</title>\n<path fill=\"none\" stroke=\"#000000\" d=\"M94,-310.8993C94,-303.0775 94,-293.9081 94,-285.5364\"/>\n<polygon fill=\"#000000\" stroke=\"#000000\" points=\"97.5001,-285.5203 94,-275.5203 90.5001,-285.5203 97.5001,-285.5203\"/>\n</g>\n<!-- gene_2_node_3 -->\n<g id=\"node18\" class=\"node\">\n<title>gene_2_node_3</title>\n<ellipse fill=\"#add8e6\" stroke=\"#000000\" cx=\"46\" cy=\"-260.2462\" rx=\"14.7\" ry=\"14.7\"/>\n<text text-anchor=\"middle\" x=\"46\" y=\"-257.1462\" font-family=\"Times,serif\" font-size=\"12.00\" fill=\"#000000\"> </text>\n</g>\n<!-- gene_2_node_0&#45;&gt;gene_2_node_3 -->\n<g id=\"edge19\" class=\"edge\">\n<title>gene_2_node_0&#45;&gt;gene_2_node_3</title>\n<path fill=\"none\" stroke=\"#000000\" d=\"M83.1081,-314.129C76.3659,-304.3391 67.6328,-291.6582 60.3326,-281.0579\"/>\n<polygon fill=\"#000000\" stroke=\"#000000\" points=\"63.1671,-279.0029 54.6126,-272.7522 57.402,-282.9732 63.1671,-279.0029\"/>\n</g>\n<!-- gene_2_node_5 -->\n<g id=\"node20\" class=\"node\">\n<title>gene_2_node_5</title>\n<ellipse fill=\"#008b00\" stroke=\"#000000\" cx=\"94\" cy=\"-128.8492\" rx=\"14.7\" ry=\"14.7\"/>\n<text text-anchor=\"middle\" x=\"94\" y=\"-125.7492\" font-family=\"Times,serif\" font-size=\"12.00\" fill=\"#000000\"> </text>\n</g>\n<!-- gene_2_node_0&#45;&gt;gene_2_node_5 -->\n<g id=\"edge22\" class=\"edge\">\n<title>gene_2_node_0&#45;&gt;gene_2_node_5</title>\n<path fill=\"none\" stroke=\"#000000\" d=\"M105.4514,-314.8109C111.3545,-306.0497 117.943,-294.522 121,-283.0955 133.7955,-235.2677 134.7363,-219.2646 121,-171.6985 118.6855,-163.6838 114.3781,-155.7245 109.8729,-148.8802\"/>\n<polygon fill=\"#000000\" stroke=\"#000000\" points=\"112.4339,-146.4436 103.7754,-140.3378 106.7364,-150.5105 112.4339,-146.4436\"/>\n</g>\n<!-- gene_2_node_2 -->\n<g id=\"node17\" class=\"node\">\n<title>gene_2_node_2</title>\n<ellipse fill=\"#add8e6\" stroke=\"#000000\" cx=\"94\" cy=\"-194.5477\" rx=\"14.7\" ry=\"14.7\"/>\n<text text-anchor=\"middle\" x=\"94\" y=\"-191.4477\" font-family=\"Times,serif\" font-size=\"12.00\" fill=\"#000000\"> </text>\n</g>\n<!-- gene_2_node_1&#45;&gt;gene_2_node_2 -->\n<g id=\"edge18\" class=\"edge\">\n<title>gene_2_node_1&#45;&gt;gene_2_node_2</title>\n<path fill=\"none\" stroke=\"#000000\" d=\"M94,-245.3382C94,-237.8847 94,-228.6536 94,-220.1139\"/>\n<polygon fill=\"#000000\" stroke=\"#000000\" points=\"97.5001,-219.8526 94,-209.8527 90.5001,-219.8527 97.5001,-219.8526\"/>\n</g>\n<!-- gene_2_node_4 -->\n<g id=\"node19\" class=\"node\">\n<title>gene_2_node_4</title>\n<ellipse fill=\"#add8e6\" stroke=\"#000000\" cx=\"46\" cy=\"-194.5477\" rx=\"14.7\" ry=\"14.7\"/>\n<text text-anchor=\"middle\" x=\"46\" y=\"-191.4477\" font-family=\"Times,serif\" font-size=\"12.00\" fill=\"#000000\"> </text>\n</g>\n<!-- gene_2_node_1&#45;&gt;gene_2_node_4 -->\n<g id=\"edge20\" class=\"edge\">\n<title>gene_2_node_1&#45;&gt;gene_2_node_4</title>\n<path fill=\"none\" stroke=\"#000000\" d=\"M85.1961,-248.1961C78.4476,-238.9594 68.9761,-225.9955 61.0343,-215.1254\"/>\n<polygon fill=\"#000000\" stroke=\"#000000\" points=\"63.7997,-212.9776 55.0743,-206.9679 58.1476,-217.1072 63.7997,-212.9776\"/>\n</g>\n<!-- gene_2_node_2&#45;&gt;gene_2_node_5 -->\n<g id=\"edge23\" class=\"edge\">\n<title>gene_2_node_2&#45;&gt;gene_2_node_5</title>\n<path fill=\"none\" stroke=\"#000000\" d=\"M94,-179.6397C94,-172.1862 94,-162.9551 94,-154.4154\"/>\n<polygon fill=\"#000000\" stroke=\"#000000\" points=\"97.5001,-154.1541 94,-144.1542 90.5001,-154.1542 97.5001,-154.1541\"/>\n</g>\n<!-- gene_2_node_3&#45;&gt;gene_2_node_4 -->\n<g id=\"edge21\" class=\"edge\">\n<title>gene_2_node_3&#45;&gt;gene_2_node_4</title>\n<path fill=\"none\" stroke=\"#000000\" d=\"M46,-245.3382C46,-237.8847 46,-228.6536 46,-220.1139\"/>\n<polygon fill=\"#000000\" stroke=\"#000000\" points=\"49.5001,-219.8526 46,-209.8527 42.5001,-219.8527 49.5001,-219.8526\"/>\n</g>\n<!-- gene_2_node_4&#45;&gt;gene_2_node_5 -->\n<g id=\"edge24\" class=\"edge\">\n<title>gene_2_node_4&#45;&gt;gene_2_node_5</title>\n<path fill=\"none\" stroke=\"#000000\" d=\"M54.8039,-182.4976C61.5524,-173.2609 71.0239,-160.297 78.9657,-149.4269\"/>\n<polygon fill=\"#000000\" stroke=\"#000000\" points=\"81.8524,-151.4087 84.9257,-141.2694 76.2003,-147.2791 81.8524,-151.4087\"/>\n</g>\n<!-- gene_2_pool -->\n<g id=\"node21\" class=\"node\">\n<title>gene_2_pool</title>\n<polygon fill=\"#ffa500\" stroke=\"#000000\" points=\"134,-78 54,-78 54,-57 134,-57 134,-78\"/>\n<text text-anchor=\"middle\" x=\"94\" y=\"-64.4\" font-family=\"Times,serif\" font-size=\"12.00\" fill=\"#000000\">Avg. Pooling</text>\n</g>\n<!-- gene_2_node_5&#45;&gt;gene_2_pool -->\n<g id=\"edge25\" class=\"edge\">\n<title>gene_2_node_5&#45;&gt;gene_2_pool</title>\n<path fill=\"none\" stroke=\"#000000\" d=\"M94,-113.9983C94,-106.3053 94,-96.7837 94,-88.3643\"/>\n<polygon fill=\"#000000\" stroke=\"#000000\" points=\"97.5001,-88.1774 94,-78.1775 90.5001,-88.1775 97.5001,-88.1774\"/>\n</g>\n<!-- linear -->\n<g id=\"node22\" class=\"node\">\n<title>linear</title>\n<polygon fill=\"#c0c0c0\" stroke=\"#000000\" points=\"121,-21 67,-21 67,0 121,0 121,-21\"/>\n<text text-anchor=\"middle\" x=\"94\" y=\"-7.4\" font-family=\"Times,serif\" font-size=\"12.00\" fill=\"#000000\">Linear</text>\n</g>\n<!-- gene_2_pool&#45;&gt;linear -->\n<g id=\"edge27\" class=\"edge\">\n<title>gene_2_pool&#45;&gt;linear</title>\n<path fill=\"none\" stroke=\"#000000\" d=\"M94,-56.7787C94,-49.6134 94,-39.9517 94,-31.3097\"/>\n<polygon fill=\"#000000\" stroke=\"#000000\" points=\"97.5001,-31.1732 94,-21.1732 90.5001,-31.1732 97.5001,-31.1732\"/>\n</g>\n</g>\n</svg>\n"
          },
          "metadata": {},
          "execution_count": 35
        }
      ]
    }
  ]
}